### ISSUE TITLE
Rare event probabilities become uncalibrated calibrated moving to .6a2 from 0.4a30 pypi version?

### ISSUE THREAD

**Issue by ryanbressler** [dmlc/xgboost#1651]
I've been working on a classifier for a rare events (~1% occurrence)  built using the "stable: 0.4a30 xgboost release from pypi.

I recently tried to upgrade it to use the new .6a2 release but found that, while roc auc didn't change much the calibration of the new models was significantly worse. 

Here are calibration plots (actual survival rates in bins of predicted survival rates) for the ne .6a2 and old .4a3 models on a holdout set:

New:
![new_calibration](https://cloud.githubusercontent.com/assets/79286/19253040/96a0c298-8f05-11e6-9d03-016b4e5e4325.png)

Old:
![old_calibration](https://cloud.githubusercontent.com/assets/79286/19253049/a3cf6f32-8f05-11e6-9b3e-0660bc2eb9e8.png)

Here is a scatter plot of the log odds predicted by each model 
![scatter_plot](https://cloud.githubusercontent.com/assets/79286/19253064/b484b800-8f05-11e6-8cee-d6391a4d832e.png)

I've repeated the experiments a number of times with different parameters/validation splits and got similar results. Notably repeated runs with .6 also seem less consistent then repeated runs with .4.

I suspect some default behavior changed that is causing this issue but i've been unable to figure out what it might be and am wondering if anyone has any thoughts?

Also thanks for xgboost!

**Comment by khotilov** [comment]
The parameter is `base_score` (global transformed bias) which is by default 0.5

**Comment by ryanbressler** [comment]
I've tried setting base_score to a variety of values and it doesn't really make a difference.

**Comment by ryanbressler** [comment]
I'm wondering specifically if there has been some change (perhaps to the pruning or regularization) that could lead to models that perform less well for these low probability events. Ie in most problems it might make sense to prune a split between 98% and 97% pure nodes it makes a big difference in this one.

I've varied and tuned the parameters a bit but am usually using 2000 trees but am usually using  something like:
    'max_depth': 4,
    'tree_method': 'exact',
    'eta': .01, 
    'gamma': 0.0,  
    'min_child_weight': 50,  
    'subsample': .5,  
    'silent': 1,
    'objective': 'binary:logistic',

Thanks again

**Comment by khotilov** [comment]
Ah, so you were already using the transformed prediction score (probability-like) to draw that log-odds plot... That's strange. I usually work with rare events and I didn't notice such calibration skew. I'll take another look.

Were your plots done using training or test data, or was there any significant overfitting?

**Comment by ryanbressler** [comment]
Thanks khotilovo. The plots are on a holdout set. This data set is quite clean and overfitting hasn't been a problem at least with that value of mid_child_weight.

I'll try a value of 0.0 for base_rate. I tried setting it to the actual rate and to .5 but it occurs that it may have been effectively 0 before the parameter was introduced?

**Comment by phunterlau** [comment]
For helping the comparison between 0.4 (Dec 2015) and 0.6 (Aug 2016), the 0.6 version on `PyPI` is github f690b2be3ae6f37390210f074cc1a84237fe4d6a from Aug 9th.

**Comment by khotilov** [comment]
For the `binary:logistic` objective, `base_score` is in the probability space, so you cannot set it to 0. Setting it to the average rate would help the booster. With the default setting and high imbalance, first trees are mostly just learning this bias.

**Comment by ryanbressler** [comment]
I've tried setting base rate to the default of .5 and to np.mean(y) but both display similar miscalibration.

**Comment by ryanbressler** [comment]
It looks like .4 was uploaded to pypi on december 8th of 2015. A lot of changes including a major refactor happened to tree learning in January and February including some things that could have had subtle effects.

The experiments I'm observing this issue in takes overnight to run but tomorrow i'm going to see about constructing a simpler experiment case and doing a binary search through the commit history to identify what caused this.

**Comment by ryanbressler** [comment]
I'm still working on this but a couple of quick updates.

A coworker noticed that if you take a model trained on .4, pickle it (python serialization for non python speakers) and reload it with .6 it will display the same miss calibration. I'm not 100% certain this is the same issue but it both suggests a bug in the prediction code and allows for quicker testing. I'll validate with full overnight retraining runs experiments once I find a likely commit.

So I searched through the tree and found the first buildable commit that shows the bias is 4a16b729fc7eb1bc81dad8bfd08f374fe7f0d2b5 from February 27th. Its a very clear cut diffrence:

7e6f00ee11cbbc361ae3d5e9162a8b6f6f11d23a	0.0
084f5f4715306f068c827d3b723b5448ee99fe7f	0.0
82ceb4de0af8877403517f5044861cf60fbd6ce8	0.0
4b4b36d047367f7417f4c9a543d9c0b18d2e4cf0	0.0
0d95e863c981548b5a7ca363310fc359a9165d85	0.0
cee148ed6435cdbbe1b03909cd47e3bdb8d73769	0.0
4a16b729fc7eb1bc81dad8bfd08f374fe7f0d2b5	-0.373062
5f70b4df7af60b9498b4c5225b8c612655a26b1e	-0.373062
803a6fe474ffe8eec7ba268220014838ebddd177	-0.373062
b7985466a4ebd5eb63037f3df710b00e3549e979	-0.373062
3d56caaab51ec5bbfc29a2c84729a6521af6411c	-0.373062

(numbers are just np.mean(loggit(predictions_from_current) - loggit(predictions_from0.4)))

Unfortunately it looks like the last buildable commit before this was cee148ed6435cdbbe1b03909cd47e3bdb8d73769 on Jan 5 leaving about 125 broken commits in the interim any one of which could have produced that change.

Spot checking these they are just failing to resolve rabit.h vs rabit/rabit.h so I'm going to see if I can patch that in my searching code and rerun that section.

Also I have some thoughts on why others may not have observed this. I'm working on a dataset with ~millions of records and where the positive events are both rare and never fully determined by the data (ie we never see more then say a 10-20% chance of seeing an event in an individual) but we care about probability accuracy for business reasons. I think its a subtle enough issue that it would be hard to notice it with a smaller data set or one where the data better determines the events or actually apply labels.

**Comment by ryanbressler** [comment]
Okay I think I may have sorted this out, i'll run a longer experiment overnight to verify but it appears to be an odd side effect of the "missing" parameter and sparse matrixes similar to https://github.com/dmlc/xgboost/issues/341

For us thing broke with f5e96eba72bde51ce7feee2437608daedbfdffc5

The thing is we're using sparse matrixes for training but not for evaluation. This was originally because scikit only supported sparse matrixes for training but dense matrixes also have some performance advantages for us when executing on small numbers of data points at a time.

It looks like the change from the default of missing=0.0 to missing=np.NaN subtly broke things for us because  calling xgb.DMatrix(X) and xgb.DMatrix(scipy.sparse.csr_matrix(X)) now produce different and apparently incompatible results.

Explicitly setting missing=0.0 fixes this for us but this is an insidious issues as it just degrades performance without causing errors. I'd suggest that some sort of runtime checks are needed.

**Comment by phunterlau** [comment]
@ryanbressler glad that you have found the reason. Thanks. As @tqchen has mentioned in https://github.com/dmlc/xgboost/issues/341#issuecomment-141797026 supporting sparse matrix naturally requires the missing value as the true missing value, in order not to confuse with 0 in sparse matrix.
